`where Self: Sized` exclusion not working with `trait_variant` support

Using cargo script for brevity:

```rust
#!/usr/bin/env -S cargo +nightly -Zscript
---cargo
[package]
edition = "2021"
[dependencies]
dynosaur = "0.1.3"
trait-variant = "0.1.2"
---

#[trait_variant::make(Run: Send)]
#[dynosaur::dynosaur(DynLocalRun = dyn LocalRun)]
#[dynosaur::dynosaur(DynRun = dyn Run)]
pub trait LocalRun {
    async fn new() -> std::io::Result<Self>
    where
        Self: Sized;

    async fn run(&mut self) -> std::io::Result<()>;
}

fn main(){}
```

Gives compile error:

```
error[E0277]: `()` is not a future
  --> /home/cdruid/.cargo/target/e4/57393f8231c91d/script.rs:11:1
   |
11 | #[trait_variant::make(Run: Send)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `()` is not a future
   |
   = help: the trait `Future` is not implemented for `()`
```

Peeking at the expansion, it looks like it's coming from:

```rust
impl<'dynosaur_struct> Run for DynRun<'dynosaur_struct> {
    fn new()
        ->
            impl ::core::future::Future<Output = std::io::Result<Self>> +
            Send where Self: Sized {
        ::core::panicking::panic("internal error: entered unreachable code")
    }
```

I think the typical workarounds here are either to use `async fn new()` (currently used for the `!Send` case) or `let fut: ::core::pin::Pin<Box<dyn ::core::future::Future<Output = std::io::Result<Self>> + 'static + Send>> = panic!(); fut` (which I believe was used in the original implementation)

--- a/dynosaur/tests/pass/where-self-sized-rpit.rs
+++ b/dynosaur/tests/pass/where-self-sized-rpit.rs
@@ -1,0 +1,11 @@
+use dynosaur::dynosaur;
+
+#[dynosaur(DynMyTrait)]
+trait MyTrait {
+    fn foo(&mut self) -> impl ::core::future::Future<Output = ()>
+    where
+        Self: Sized;
+    async fn bar(&mut self);
+}
+
+fn main() {}

--- a/dynosaur/tests/pass/where-self-sized-rpit.stdout
+++ b/dynosaur/tests/pass/where-self-sized-rpit.stdout
@@ -1,0 +1,80 @@
+#![feature(prelude_import)]
+#[prelude_import]
+use std::prelude::rust_2021::*;
+#[macro_use]
+extern crate std;
+use dynosaur::dynosaur;
+
+trait MyTrait {
+    fn foo(&mut self)
+    -> impl ::core::future::Future<Output = ()>
+    where
+    Self: Sized;
+    async fn bar(&mut self);
+}
+mod __dynosaur_macro_dynmytrait {
+    use super::*;
+    trait ErasedMyTrait {
+        fn bar<'life0, 'dynosaur>(&'life0 mut self)
+        ->
+            ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
+            'dynosaur>>
+        where
+        'life0: 'dynosaur,
+        Self: 'dynosaur;
+    }
+    impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
+        fn bar<'life0, 'dynosaur>(&'life0 mut self)
+            ->
+                ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
+                'dynosaur>> where 'life0: 'dynosaur, Self: 'dynosaur {
+            Box::pin(<Self as MyTrait>::bar(self))
+        }
+    }
+    #[repr(transparent)]
+    pub struct DynMyTrait<'dynosaur_struct> {
+        ptr: dyn ErasedMyTrait + 'dynosaur_struct,
+    }
+    impl<'dynosaur_struct> MyTrait for DynMyTrait<'dynosaur_struct> {
+        #[allow(unreachable_code)]
+        fn foo(&mut self) -> impl ::core::future::Future<Output = ()> where
+            Self: Sized {
+
+
+                ::core::panicking::panic("internal error: entered unreachable code")
+                as
+                ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
+                'static>>
+        }
+        fn bar(&mut self) -> impl ::core::future::Future<Output = ()> {
+            let ret:
+                    ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
+                    ()> + '_>> = self.ptr.bar();
+            let ret:
+                    ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
+                    ()> + 'static>> = unsafe { ::core::mem::transmute(ret) };
+            ret
+        }
+    }
+    impl<'dynosaur_struct> DynMyTrait<'dynosaur_struct> {
+        pub fn boxed(value: impl MyTrait + 'dynosaur_struct)
+            -> Box<DynMyTrait<'dynosaur_struct>> {
+            let value = Box::new(value);
+            let value: Box<dyn ErasedMyTrait + 'dynosaur_struct> = value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub fn from_ref(value: &(impl MyTrait + 'dynosaur_struct))
+            -> &DynMyTrait<'dynosaur_struct> {
+            let value: &(dyn ErasedMyTrait + 'dynosaur_struct) = &*value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub fn from_mut(value: &mut (impl MyTrait + 'dynosaur_struct))
+            -> &mut DynMyTrait<'dynosaur_struct> {
+            let value: &mut (dyn ErasedMyTrait + 'dynosaur_struct) =
+                &mut *value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+    }
+}
+use __dynosaur_macro_dynmytrait::DynMyTrait;
+fn main() {}

--- a/dynosaur_derive/src/expand.rs
+++ b/dynosaur_derive/src/expand.rs
@@ -276,10 +276,20 @@ pub(crate) fn expand_blanket_impl_fn(item_trait: &ItemTrait, sig: &mut Signature
 }
 
 pub(crate) fn expand_dyn_struct_fn(sig: &Signature) -> TokenStream {
-    if has_where_self_sized(&sig) {
-        quote! {
-            #sig {
-                unreachable!()
+    if has_where_self_sized(sig) {
+        if is_rpit(sig) {
+            let ty = expand_sig_ret_ty(&sig, "'static");
+            quote! {
+                #[allow(unreachable_code)]
+                #sig {
+                    unreachable!() as #ty
+                }
+            }
+        } else {
+            quote! {
+                #sig {
+                    unreachable!()
+                }
             }
         }
     } else {

--- a/dynosaur_derive/src/sig.rs
+++ b/dynosaur_derive/src/sig.rs
@@ -38,7 +38,7 @@ pub(crate) fn is_rpit(sig: &Signature) -> bool {
     match sig {
         Signature {
             asyncness: Some(_), ..
-        } => true,
+        } => false,
         Signature {
             asyncness: None,
             output: ReturnType::Type(_, ret),

